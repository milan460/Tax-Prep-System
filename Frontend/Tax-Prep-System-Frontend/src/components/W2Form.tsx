import React, { useState, useEffect } from "react";
import { Form, GridContainer, Grid } from '@trussworks/react-uswds';

interface W2FormData {
    id?: string;
    income: string;
    socialSecurityWages: string;
    medicareWages: string;
    socialSecurityTaxWithheld: string;
    medicareTaxWithheld: string;
    federalIncomeWithheld: string;
    w2StreetAddress1: string;
    w2StreetAddress2: string;
    w2City: string;
    w2State: string;
    w2Zip: string;
}

interface User {
    userId: number;
    username: string;
    password: string;
    role: string;
}

export default function W2Form() {
    const initialFormData: W2FormData = {
        income: "",
        socialSecurityWages: "",
        medicareWages: "",
        socialSecurityTaxWithheld: "",
        medicareTaxWithheld: "",
        federalIncomeWithheld: "",
        w2StreetAddress1: "",
        w2StreetAddress2: "",
        w2City: "",
        w2State: "",
        w2Zip: "",
    };

    const initialUser: User = {
        userId: 4,
        username: "aatrox",
        password: "bxa(n`0Xqujjd~",
        role: "ROLE_USER"
    };

    const [formData, setFormData] = useState<W2FormData>(initialFormData);
    const [formSubmitted, setFormSubmitted] = useState<boolean>(false);

    useEffect(() => {
        const savedFormData = localStorage.getItem('w2FormData');
        if (savedFormData) {
            setFormData(JSON.parse(savedFormData));
        }

        const formSubmittedStatus = localStorage.getItem(`formSubmitted_${initialUser.userId}`);
        if (formSubmittedStatus === 'true') {
            setFormSubmitted(true);
        }
    }, [initialUser.userId]);

    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = e.target;
        setFormData((prevData) => ({
            ...prevData,
            [name]: value
        }));
    };

    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
    
        const userId: number = initialUser.userId;
    
        let requestOptions: any;
        const formDataToSend = {
            user: {
                userId: initialUser.userId
            },
            income: parseFloat(formData.income),
            socialSecurityWages: parseFloat(formData.socialSecurityWages),
            medicareWages: parseFloat(formData.medicareWages),
            socialSecurityTaxWithheld: parseFloat(formData.socialSecurityTaxWithheld),
            medicareTaxWithheld: parseFloat(formData.medicareTaxWithheld),
            federalIncomeWithheld: parseFloat(formData.federalIncomeWithheld),
            streetAddress1: formData.w2StreetAddress1,
            streetAddress2: formData.w2StreetAddress2,
            city: formData.w2City,
            state: formData.w2State,
            zip: formData.w2Zip
        };
    
        if (!formSubmitted) {
            requestOptions = {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(formDataToSend)
            };
            setFormSubmitted(true);
            localStorage.setItem(`formSubmitted_${userId}`, 'true');
        } else {
            const formDataToSendWithId = { ...formDataToSend, id: userId.toString() };
            requestOptions = {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(formDataToSendWithId)
            };
        }
    
        try {
            let url = 'http://localhost:8080/w2Forms/createW2Form';
            if (formSubmitted) {
                url = `http://localhost:8080/w2Forms/updateW2Form/${userId}`;
            }
    
            const response = await fetch(url, requestOptions);
            if (!response.ok) {
                throw new Error('Network response was not ok');
            }
            if (response.status === 201) {
                const formId = response.headers.get('Location');
                if (formId) {
                    localStorage.setItem('formId', formId);
                }
            }
        } catch (error) {
            console.error('There was a problem with the fetch operation:', error);
        }
    
        localStorage.setItem('w2FormData', JSON.stringify(formData));
    };

    return (
        <>
            <GridContainer>
                <Grid row>
                    <div className="usa-layout-docs__main desktop:grid-col-12 tablet:grid-col-8">
                        <Form onSubmit={handleSubmit}>
                            {Object.entries(formData).map(([key, value]) => (
                                <div key={key} className="usa-form-group">
                                    <label htmlFor={key}>{key}:</label>
                                    <input
                                        id={key}
                                        type="text"
                                        name={key}
                                        value={value}
                                        onChange={handleChange}
                                        className="usa-input"
                                    />
                                </div>
                            ))}
                            <button type="submit" className="usa-button">Submit</button>
                        </Form>
                    </div>
                </Grid>
            </GridContainer>
        </>
    );
}