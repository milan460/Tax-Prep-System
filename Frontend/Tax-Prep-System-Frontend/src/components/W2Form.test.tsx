import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import W2Form from './W2Form';
import fetchMock from 'jest-fetch-mock';

// Mock the useNavigate hook from react-router-dom using Jest's jest.mock function
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn()
}));

// Define a MockRouter component using MemoryRouter from react-router-dom
const MockRouter: React.FC<{ children: React.ReactNode }> = ({ children }) => (
    <MemoryRouter initialEntries={['/']}>{children}</MemoryRouter>
);

// Snapshot test to ensure W2Form component renders correctly
it('renders correctly', () => {
    const setCurrentPageMock = jest.fn();

    // Render the W2Form component wrapped in BrowserRouter for testing
    const component = renderer.create(
        <BrowserRouter>
            <W2Form setCurrentPage={setCurrentPageMock} />
        </BrowserRouter>
    );

    // Generate a snapshot and compare with the previous snapshot
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

// Component tests for the W2Form component
describe('W2Form component', () => {
    // Reset fetch mocks before each test
    beforeEach(() => {
        fetchMock.resetMocks();
    });

    // Test if the component fetches the current user data successfully
    test('fetches current user successfully', async () => {
        // Mock the fetch response with user data
        fetchMock.mockResponseOnce(JSON.stringify({ userId: 1, username: 'testUser' }));

        // Render the W2Form component wrapped in MockRouter for testing
        render(
            <MockRouter>
                <W2Form setCurrentPage={jest.fn()} />
            </MockRouter>
        );
    });

    // Test if the component handles fetch errors correctly
    test('handles fetch error', async () => {
        // Mock the fetch rejection with an error
        fetchMock.mockRejectOnce(new Error('Fetch error'));

        // Render the W2Form component wrapped in MockRouter for testing
        render(
            <MockRouter>
                <W2Form setCurrentPage={jest.fn()} />
            </MockRouter>
        );
    });

    // Test if the component updates the formData state correctly on input change
    test('updates formData state on input change', () => {
        // Render the W2Form component wrapped in MockRouter for testing
        const { container } = render(
            <MockRouter>
                <W2Form setCurrentPage={jest.fn()} />
            </MockRouter>
        );
        
        // Simulate an input change and test if the state is updated correctly
        const inputElement = container.querySelector('#income') as HTMLInputElement;
        fireEvent.change(inputElement, { target: { name: 'income', value: '50000' } });
        expect(inputElement.value).toBe('50000');
    });

    // Test if the component updates the formData state correctly on state change
    test('updates formData state on state change', () => {
        // Render the W2Form component without MockRouter for testing
        const { container } = render(<W2Form setCurrentPage={jest.fn()} />);
        
        // Simulate a select change and test if the state is updated correctly
        const selectElement = container.querySelector('#state') as HTMLSelectElement;
        fireEvent.change(selectElement, { target: { value: 'California' } });
        expect(selectElement.value).toBe('California');
    });
});