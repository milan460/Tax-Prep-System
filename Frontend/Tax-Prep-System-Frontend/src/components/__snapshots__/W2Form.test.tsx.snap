// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1>
    w2FormHeading
  </h1>
  <div
    className="grid-container"
    data-testid="gridContainer"
  >
    <div
      className="grid-row"
      data-testid="grid"
    >
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="income"
        >
          incomeLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="income"
          name="income"
          onChange={[Function]}
          placeholder="incomePlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="number"
          value=""
        />
      </div>
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="federalIncomeWithheld"
        >
          federalIncomeWithheldLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="federalIncomeWithheld"
          name="federalIncomeWithheld"
          onChange={[Function]}
          placeholder="federalIncomeWithheldPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="number"
          value=""
        />
      </div>
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="medicareWages"
        >
          medicareWagesLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="medicareWages"
          name="medicareWages"
          onChange={[Function]}
          placeholder="medicareWagesPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="number"
          value=""
        />
      </div>
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="medicareTaxWithheld"
        >
          medicareTaxWithheldLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="medicareTaxWithheld"
          name="medicareTaxWithheld"
          onChange={[Function]}
          placeholder="medicareTaxWithheldPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="number"
          value=""
        />
      </div>
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="socialSecurityWages"
        >
          socialSecurityWagesLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="socialSecurityWages"
          name="socialSecurityWages"
          onChange={[Function]}
          placeholder="socialSecurityWagesPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="number"
          value=""
        />
      </div>
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="socialSecurityTaxWithheld"
        >
          socialSecurityTaxWithheldLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="socialSecurityTaxWithheld"
          name="socialSecurityTaxWithheld"
          onChange={[Function]}
          placeholder="socialSecurityTaxWithheldPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="number"
          value=""
        />
      </div>
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="streetAddress1"
        >
          employerStreetAddress1Label
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="streetAddress1"
          name="streetAddress1"
          onChange={[Function]}
          placeholder="employerStreetAddress1Placeholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="grid-col-4"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="streetAddress2"
        >
          employerStreetAddress2Label
           
          <span
            className="text-italic"
          >
            - 
            optional
          </span>
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="streetAddress2"
          name="streetAddress2"
          onChange={[Function]}
          placeholder="employerStreetAddress2Placeholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-2"
        data-testid="grid"
      />
      <div
        className="grid-col-3"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="city"
        >
          employerCityLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="city"
          name="city"
          onChange={[Function]}
          placeholder="employerCityPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="grid-col-3 usa-form-group"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="state"
        >
          employerStateLabel
        </label>
        <select
          className="usa-select"
          data-testid="Select"
          id="state"
          name="state"
          onChange={[Function]}
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          value=""
        >
          <option
            value=""
          >
            statePlaceholder
          </option>
          <option
            value="Alabama"
          >
            alabama
          </option>
          <option
            value="Alaska"
          >
            alaska
          </option>
          <option
            value="Arizona"
          >
            arizona
          </option>
          <option
            value="Arkansas"
          >
            arkansas
          </option>
          <option
            value="California"
          >
            california
          </option>
          <option
            value="Colorado"
          >
            colorado
          </option>
          <option
            value="Connecticut"
          >
            connecticut
          </option>
          <option
            value="Delaware"
          >
            delaware
          </option>
          <option
            value="District Of Columbia"
          >
            districtOfColumbia
          </option>
          <option
            value="Florida"
          >
            florida
          </option>
          <option
            value="Georgia"
          >
            georgia
          </option>
          <option
            value="Hawaii"
          >
            hawaii
          </option>
          <option
            value="Idaho"
          >
            idaho
          </option>
          <option
            value="Illinois"
          >
            illinois
          </option>
          <option
            value="Indiana"
          >
            indiana
          </option>
          <option
            value="Iowa"
          >
            iowa
          </option>
          <option
            value="Kansas"
          >
            kansas
          </option>
          <option
            value="Kentucky"
          >
            kentucky
          </option>
          <option
            value="Louisiana"
          >
            louisiana
          </option>
          <option
            value="Maine"
          >
            maine
          </option>
          <option
            value="Maryland"
          >
            maryland
          </option>
          <option
            value="Massachusetts"
          >
            massachusetts
          </option>
          <option
            value="Michigan"
          >
            michigan
          </option>
          <option
            value="Minnesota"
          >
            minnesota
          </option>
          <option
            value="Mississippi"
          >
            mississippi
          </option>
          <option
            value="Missouri"
          >
            missouri
          </option>
          <option
            value="Montana"
          >
            montana
          </option>
          <option
            value="Nebraska"
          >
            nebraska
          </option>
          <option
            value="Nevada"
          >
            nevada
          </option>
          <option
            value="New Hampshire"
          >
            newHampshire
          </option>
          <option
            value="New Jersey"
          >
            newJersey
          </option>
          <option
            value="New Mexico"
          >
            newMexico
          </option>
          <option
            value="New York"
          >
            newYork
          </option>
          <option
            value="North Carolina"
          >
            northCarolina
          </option>
          <option
            value="North Dakota"
          >
            northDakota
          </option>
          <option
            value="Ohio"
          >
            ohio
          </option>
          <option
            value="Oklahoma"
          >
            oklahoma
          </option>
          <option
            value="Oregon"
          >
            oregon
          </option>
          <option
            value="Pennsylvania"
          >
            pennsylvania
          </option>
          <option
            value="Rhode Island"
          >
            rhodeIsland
          </option>
          <option
            value="South Carolina"
          >
            southCarolina
          </option>
          <option
            value="South Dakota"
          >
            southDakota
          </option>
          <option
            value="Tennessee"
          >
            tennessee
          </option>
          <option
            value="Texas"
          >
            texas
          </option>
          <option
            value="Utah"
          >
            utah
          </option>
          <option
            value="Vermont"
          >
            vermont
          </option>
          <option
            value="Virginia"
          >
            virginia
          </option>
          <option
            value="Washington"
          >
            washington
          </option>
          <option
            value="West Virginia"
          >
            westVirginia
          </option>
          <option
            value="Wisconsin"
          >
            wisconsin
          </option>
          <option
            value="Wyoming"
          >
            wyoming
          </option>
        </select>
      </div>
      <div
        className="grid-col-2"
        data-testid="grid"
      >
        <label
          className="usa-label text-bold text-underline text-info-darker"
          data-testid="label"
          htmlFor="zip"
        >
          zipLabel
        </label>
        <input
          className="usa-input"
          data-testid="textInput"
          id="zip"
          name="zip"
          onChange={[Function]}
          placeholder="zipPlaceholder"
          style={
            {
              "width": "calc(100% - 16px)",
            }
          }
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="grid-container"
    data-testid="gridContainer"
  >
    <div
      className="grid-row"
      data-testid="grid"
    >
      <div
        className="grid-col-12"
        data-testid="grid"
      >
        <button
          className="usa-button usa-button--base"
          data-testid="button"
          onClick={[Function]}
          style={
            {
              "marginBottom": "20px",
              "marginTop": "20px",
            }
          }
          type="button"
        >
          backButton
        </button>
        <button
          className="usa-button"
          data-testid="button"
          onClick={[Function]}
          type="button"
        >
          continueButton
        </button>
      </div>
    </div>
  </div>
</div>
`;
